platform-intel.c

```diff
@@ -669,14 +724,251 @@ static int read_efi_variable(void *buffer, ssize_t buf_size,
+ * If found, the buffer with the OROM structure will be filled.
+ *
+ * Return: %MDADM_STATUS_SUCCESS on success, %MDADM_STATUS_ERROR otherwise.
+ */
+mdadm_status_t
+find_orom_in_acpi_uefi_tables(uefi_controller_t *uefi_controllers, int uefi_controllers_number,
+			      struct imsm_orom *orom)
+{
+	mdadm_status_t status = MDADM_STATUS_ERROR;
+	uefi_acpi_table_with_orom_t uefi_table;
+	char path[PATH_MAX];
+	struct dirent *ent;
+	int fd = -1;
+	DIR *dir;
+
+	dir = opendir(ACPI_TABLES_PATH);
+	if (!dir)
+		return MDADM_STATUS_ERROR;
+
+	for (ent = readdir(dir); ent; ent = readdir(dir)) {
+		close_fd(&fd);
+
+		/* Check if file is a UEFI table */
+		if (strncmp(ent->d_name, ACPI_UEFI_TABLE_BASE_NAME,
+			    strlen(ACPI_UEFI_TABLE_BASE_NAME)) != 0)
+			continue;
+
+		snprintf(path, PATH_MAX, "%s/%s", ACPI_TABLES_PATH, ent->d_name);
+
+		fd = open(path, O_RDONLY);
+		if (!is_fd_valid(fd)) {
+			pr_err("Fail to open ACPI UEFI table file. File: %s, Error: %s\n",
+			       ent->d_name, strerror(errno));
+			continue;
+		}
+
+		if (read_uefi_acpi_table(fd, ent->d_name, &uefi_table.table))
+			continue;
+
+		if (!acpi_is_requested_uefi_controller(uefi_controllers, uefi_controllers_number,
+						       uefi_table.table))
+			continue;
+
+		if (!verify_uefi_acpi_table_length(uefi_table.table))
+			continue;
+
+		if (read_uefi_acpi_orom_data(fd, ent->d_name, &uefi_table)) {
+			continue;
+		} else {
+			memcpy(orom, &uefi_table.orom, sizeof(uefi_table.orom));
+			status = MDADM_STATUS_SUCCESS;
+			goto close;
+		}
+	}
+
+close:
+	close_fd(&fd);
+	closedir(dir);
+	return status;
+}
+
+/**
+ * find_imsm_efi_orom() - find OROM for requested controller.
+ * @orom: buffer for OROM.
+ * @controller_type: requested controller type.
+ *
+ * Based on controller type, function first search in EFI vars then in ACPI UEFI tables.
+ * In case of success, the structure &orom will be filed out.
+ *
+ * Return: %MDADM_STATUS_SUCCESS on success.
+ */
+static mdadm_status_t
+find_imsm_efi_orom(struct imsm_orom *orom, enum sys_dev_type controller_type)
+{
+	static uefi_controller_t sata_uefi_controllers[] = {
+		{AHCI_PROP, RST_SATA_V_GUID},
+		{AHCI_SSATA_PROP, RST_SSATA_V_GUID},
+		{AHCI_TSATA_PROP, RST_TSATA_V_GUID},
+	};
+	static uefi_controller_t vmd_uefi_controllers[] = {
+		{VROC_VMD_PROP, RST_UEFI_V_GUID},
+		{RST_VMD_PROP, RST_VMD_V_GUID},
+	};
+	static uefi_controller_t *uefi_controllers;
+	int uefi_controllers_number;
+	int index;
+
+	switch (controller_type) {
+	case SYS_DEV_SATA:
+		uefi_controllers = sata_uefi_controllers;
+		uefi_controllers_number = ARRAY_SIZE(sata_uefi_controllers);
+		break;
+	case SYS_DEV_VMD:
+	case SYS_DEV_SATA_VMD:
+		uefi_controllers = vmd_uefi_controllers;
+		uefi_controllers_number = ARRAY_SIZE(vmd_uefi_controllers);
+		break;
+	default:
+		return MDADM_STATUS_UNDEF;
+	}
+
+	/* Attempt to read via EFI vars. */
+	for (index = 0; index < uefi_controllers_number; index++)
```

Comment:
For readability, please move this for to the function. It got me confused because once we have a loop and later we are using function. My confusion probably comes from first match wins conception used here.

```c
if (find_orom_in_efi_var_fs(uefi_controllers, uefi_controllers_number, orom);
	return MDADM_STATUS_SUCCESS;
return find_orom_in_acpi_uefi_tables(uefi_controllers, uefi_controllers_number, orom);
```